### PR TITLE
[ base ] Some properties of `List` and `SnocList`

### PR DESCRIPTION
**Pull request by buzden**
...or else it is hard to write even simple properties involving blend of `List1`, `SnocList` and `List`.

### CODE REVIEW

File: libs/base/Data/List.idr [hunk 1]
```diff
@@ -1027,3 +1027,8 @@ export
 lengthReplicate : (n : Nat) -> length (replicate n x) = n
 lengthReplicate 0 = Refl
 lengthReplicate (S k) = cong S (lengthReplicate k)
+
+export
+foldlOfAppend : (xs : List a) -> (ys : List a) -> (f : acc -> a -> acc) -> (init : acc) -> foldl f init (xs ++ ys) = foldl f (foldl f init xs) ys
```
Comment (gallais) [inline libs/base/Data/List.idr:1032]: Please use the same naming scheme as in the rest of the file:
* lemma of the shape `fun1 (fun2 xs) = ...` is named `fun1Fun2`
* no `Of` in the middle of the name
* arguments taken in the same left-to-right order they're in in the statement, here: `f, init, xs, ys

File: libs/base/Data/SnocList.idr [hunk 2]
```diff
@@ -138,3 +138,50 @@ findIndex _ Lin = Nothing
...
+
+export
+appendLinLeftNeutral : (sx : SnocList a) -> [<] ++ sx = sx
+appendLinLeftNeutral [<]       = Refl
+appendLinLeftNeutral (sx :< _) = rewrite appendLinLeftNeutral sx in Refl
+
+--- Append of snoc-list and usual list ---
+
+export
+heterListAppendAsListAppend : (xs : SnocList a) -> (ys : List a) -> xs <>> ys = cast xs ++ ys
```
Comment (gallais) [inline libs/base/Data/SnocList.idr:161]: This `(<>>)` combinator is called chips

File: libs/base/Data/SnocList.idr [hunk 3]
```diff
@@ -138,3 +138,50 @@ findIndex _ Lin = Nothing
...
+heterListAppendAsListAppend [<]       ys = Refl
+heterListAppendAsListAppend (sx :< x) ys = do
+  rewrite heterListAppendAsListAppend sx (x :: ys)
+  rewrite heterListAppendAsListAppend sx [x]
+  rewrite sym $ appendAssociative (cast sx) [x] ys
+  Refl
+
+export
+heterSnocAppendAsSnocAppend : (xs : SnocList a) -> (ys : List a) -> xs <>< ys = xs ++ cast ys
```
Comment (gallais) [inline libs/base/Data/SnocList.idr:170]: And this one `(<><)` is fish

**Comment by buzden** [comment]
Hopefully I resolved the comments, rebased and force-pushed. Also, I've added a couple more properties, I hope that's okay.

File: libs/base/Data/List.idr (hunk 1 above)
Comment (buzden) [reply libs/base/Data/List.idr]: I wonder why `++` is already called `append`. Isn't it a concatenation, when appending is what `snoc` function does?

**Pull request merged**